### ISSUE TITLE
Optimize loading of MP4 on slow media sources

### ISSUE THREAD

**Issue by pdeljanov** [pdeljanov/Symphonia#175]
The initial scan of top-level atoms in `IsoMp4Reader::try_new` can be slow if the `MediaSource` is slow to read and seek. This is particularly bad with fragmented MP4s where there are many MOOF + MDAT pairs. This issue tracks possible optimizations:

1. `AtomIterator::next` calls `ignore_bytes` to skip the body of an unread atom. It should be possible to replace `ignore_bytes` with an explicit `seek` if the reader is seekable. However, there will still be a lot of seek + small read operations as all the atoms are skipped over.
2. If the MOOV and FTYP atoms were already encountered, it may be reasonable to terminate scanning of the top-level atoms at the first MOOF or MDAT atom. There'd need to be some documentation (i.e., standards, etc.) or proof that this is a good general optimization that won't break anything. However, this would still defer loading the fragments to the first seek.
3. Use the SIDX atom (segment index) to avoid the scanning that would happen on a seek if the 2nd point was implemented.

**Comment by erikas-taroza** [comment]
From #153

> If I understand correctly, it sounds like the MediaSource works properly now and the only problem is that loading the MP4 is slow?

Streaming a normal MP4 works fine. The loading of a fragmented MP4 is slow because of the chunk requests.

> However, there will still be a lot of seek + small read operations as all the atoms are skipped over..

I believe the atom skipping is what is causing the HTTP requests to happen, making the loading slow. Therefore, I do not think solution 1 will help (in the case of streaming a fragmented MP4).

**Comment by pdeljanov** [comment]
> > However, there will still be a lot of seek + small read operations as all the atoms are skipped over..
> 
> I believe the atom skipping is what is causing the HTTP requests to happen, making the loading slow. Therefore, I do not think solution 1 will help (in the case of streaming a fragmented MP4).

Probably not by much.

You can try this to see if optimization 2 helps. In this match case:

https://github.com/pdeljanov/Symphonia/blob/412f44daab39920beeb81d78b0e4271b263d33e9/symphonia-format-isomp4/src/demuxer.rs#L384-L398

Replace the contents with:

```rust
if moov.is_some() && ftyp.is_some() {
    break;
}
```

Loading should be much faster.

**Comment by erikas-taroza** [comment]
With that change, initial loading does not download the whole file (good!). When seeking, `ignore_bytes` is called and it reads until the seek position is found.

**Comment by pdeljanov** [comment]
> With that change, initial loading does not download the whole file (good!). When seeking, `ignore_bytes` is called and it reads until the seek position is found.

Yes, as mentioned, this defers further scanning to when it seeks. However, you only have to scan from current position to the required position so still a net positive. If implemented, the 1st item should speed the scan up further if you want to try. Of course, the 3rd item would be best, but that's a non-trivial change to even prototype.

**Comment by erikas-taroza** [comment]
> If implemented, the 1st item should speed the scan up further if you want to try.

I can give it a go.

**Comment by erikas-taroza** [comment]
Hello @pdeljanov 

[This](https://github.com/erikas-taroza/Symphonia/commit/08ec09acd047a1cb3fe5c9a04f306f97b6b8e3c9) is what I got. Let me know if you have any suggestions with the way I implemented it.

I used `std::time::Instant` to test the difference. Seeking takes 0ms and calling `ignore_bytes` takes about 300-400ms. It was taking around 75-80ms every time it was skipping.

**Comment by pdeljanov** [comment]
I don't think this is how we would implement it since `ReadBytes` shouldn't have any seek functionality (it's a stream), but it seems reasonable for experimentation. Does this solve the problem for you?

**Comment by erikas-taroza** [comment]
> I don't think this is how we would implement it since `ReadBytes` shouldn't have any seek functionality (it's a stream), but it seems reasonable for experimentation. Does this solve the problem for you?

Yeah its a naive implementation. It doesn't completely solve the problem though. I think using the SIDX atom would be the final step. Unfortunately, I do not know how to implement it.

**Comment by pdeljanov** [comment]
Yeah, so to recap: the file loads instantly now, but you're seeing multiple seeks as we jump from fragment to fragment until the timestamp is reached?

The SIDX atom would help here I think. Though, it is hard to implement and may raise other issues during implementation. Unfortunately, it will take some time.

Another thing you could do right now is reuse your HTTP connection(s). I think a lot of time is likely being spent setting up a new connection (especially with SSL) so each seek becomes very expensive. Better yet, if QUIC is offered by the server, it'll be even faster.

**Comment by erikas-taroza** [comment]
> Yeah, so to recap: the file loads instantly now, but you're seeing multiple seeks as we jump from fragment to fragment until the timestamp is reached?

Yes, this is the current behavior.

> Another thing you could do right now is reuse your HTTP connection(s).

I will look into this.
